clients/geth/specular/entry/geth/hook.go

```diff
@@ -0,0 +1,154 @@
+
+type RollupConfig interface {
+	GetCoinbase() common.Address        // recipient of the L1 Fee
+	GetL2ChainID() uint64               // chain ID of the specular rollup
+	GetL1FeeOverhead() int64            // fixed cost of submitting a tx to L1
+	GetL1FeeMultiplier() float64        // value to scale the L1 Fee
+	GetL1OracleAddress() common.Address // contract providing the L1 basefee
+}
+
+// SpecularEVMPreTransferHook is injected into the EVM and runs before every transfer
+// currently this is only used to calculate & charge the L1 Fee
+func SpecularEVMPreTransferHook(msg types.Message, evm *vm.EVM, cfg RollupConfig) error {
+	tx := transactionFromMessage(msg, cfg)
+	fee, err := calculateL1Fee(tx, evm, cfg)
+	if err != nil {
+		return err
+	}
+	return chargeL1Fee(fee, msg, evm, cfg)
+}
+
+// creates a Transaction from a transaction
+// the Tx Type is reconstructed and the signature is left empty
+func transactionFromMessage(msg types.Message, cfg RollupConfig) *types.Transaction {
+	var txData types.TxData
+	if msg.GasTipCap() != nil {
+		txData = &types.DynamicFeeTx{
+			ChainID:    big.NewInt(int64(cfg.GetL2ChainID())),
+			Nonce:      msg.Nonce(),
+			GasTipCap:  msg.GasTipCap(),
+			GasFeeCap:  msg.GasFeeCap(),
+			Gas:        msg.Gas(),
+			To:         msg.To(),
+			Value:      msg.Value(),
+			Data:       msg.Data(),
+			AccessList: msg.AccessList(),
+		}
+	} else if msg.AccessList() != nil {
+		txData = &types.AccessListTx{
+			ChainID:    big.NewInt(int64(cfg.GetL2ChainID())),
+			Nonce:      msg.Nonce(),
+			GasPrice:   msg.GasPrice(),
+			Gas:        msg.Gas(),
+			To:         msg.To(),
+			Value:      msg.Value(),
+			Data:       msg.Data(),
+			AccessList: msg.AccessList(),
+		}
+	} else {
+		txData = &types.LegacyTx{
+			Nonce:    msg.Nonce(),
+			GasPrice: msg.GasPrice(),
+			Gas:      msg.Gas(),
+			To:       msg.To(),
+			Value:    msg.Value(),
+			Data:     msg.Data(),
+		}
+	}
+	return types.NewTx(txData)
+}
+
+// calculates the L1 fee for a transaction using the following formula:
+// L1Fee = L1FeeMultiplier * L1BaseFee * (TxDataGas + L1OverheadGas)
+// L1BaseFee is dynamically set by the L1Oracle
+// L1FeeMultiplier & L1OverheadGas are set in the rollup configuration
+func calculateL1Fee(tx *types.Transaction, evm *vm.EVM, cfg RollupConfig) (*big.Int, error) {
+	// calculate L1 gas from RLP encoding
+	buf := new(bytes.Buffer)
+	err := tx.EncodeRLP(buf)
+	if err != nil {
+		return common.Big0, err
+	}
+	bytes := buf.Bytes()
+	// remove the last 3 bytes containing the signature
+	// this mirrors the optimism implementation [1]
+	// but contradicts the optimism spec [2]
+	// [1] https://github.com/ethereum-optimism/optimism/blob/5d9a38dcd9dc79dce41a6d08f9b28ff850f77811/l2geth/rollup/fees/rollup_fee.go#L204
+	// [2] https://github.com/ethereum-optimism/optimism/blob/develop/specs/exec-engine.md#l1-cost-fees-l1-fee-vault
+	rlp := bytes[:len(bytes)-3]
```

Comment:
can you elaborate a bit, isn't the signature 65B

Reply:
since the tx is unsigned we have `r, s, v = 0` stored as `big.Int`. when RLP encoding geth encodes `big.Int` as integers ([docs](https://github.com/SpecularL2/go-ethereum/blob/e5b0d5f8cf3f7b8c0c20602ffd5a0199c342d4c2/rlp/doc.go#L52)) which leads to all 3 values getting serialised to 0 bytes